cuda/factorization/par_ilu_kernels.cu

```diff
@@ -48,33 +54,218 @@ namespace cuda {
 namespace par_ilu_factorization {
 
 
+constexpr int default_block_size{512};
+
+
+namespace kernel {
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void count_nnz_per_l_u_row(
+    size_type num_rows, const IndexType *__restrict__ row_ptrs,
+    const IndexType *__restrict__ col_idxs,
+    const ValueType *__restrict__ values, IndexType *__restrict__ l_nnz_row,
+    IndexType *__restrict__ u_nnz_row)
+{
+    const auto row = blockDim.x * blockIdx.x + threadIdx.x;
+    if (row < num_rows) {
+        IndexType l_row_nnz{};
+        IndexType u_row_nnz{};
+        for (auto idx = row_ptrs[row]; idx < row_ptrs[row + 1]; ++idx) {
+            auto col = col_idxs[idx];
+            l_row_nnz += (col <= row);
+            u_row_nnz += (row <= col);
+        }
+        l_nnz_row[row] = l_row_nnz;
+        u_nnz_row[row] = u_row_nnz;
+    }
+}
+
+
+}  // namespace kernel
+
+
 template <typename ValueType, typename IndexType>
 void initialize_row_ptrs_l_u(
     std::shared_ptr<const CudaExecutor> exec,
     const matrix::Csr<ValueType, IndexType> *system_matrix,
-    IndexType *l_row_ptrs, IndexType *u_row_ptrs) GKO_NOT_IMPLEMENTED;
+    IndexType *l_row_ptrs, IndexType *u_row_ptrs)
+{
+    const size_type num_rows{system_matrix->get_size()[0]};
+    const size_type num_row_ptrs{num_rows + 1};
+
+    const dim3 block_size{default_block_size, 1, 1};
+    const dim3 grid_dim{static_cast<uint32>(ceildiv(
+                            num_rows, static_cast<size_type>(block_size.x))),
```

Comment:
You could store the `ceildiv` result in an `int`. This would remove the `static_cast<uint32>` and maybe increase a bit readability.